### PR TITLE
more strict on nulls, ::class over string-class, early returns

### PR DESCRIPTION
**Pull request by inoas**
I hope I got the (hopefully) better parts of https://github.com/cakephp/authentication/pull/175 isolated.
I still have a very bad feeling about `isset()` and `empty()` TBH and I feel things my gut feeling is that things might bite back in 7.1/7.2.

### CODE REVIEW

File: src/Authenticator/HttpDigestAuthenticator.php [hunk 1]
```diff
@@ -114,7 +114,7 @@ public function authenticate(ServerRequestInterface $request, ResponseInterface
      * Gets the digest headers from the request/environment.
      *
      * @param \Psr\Http\Message\ServerRequestInterface $request The request that contains login information.
-     * @return array Array of digest information.
+     * @return array|null Array of digest information.
```
Comment (dereuromark) [inline src/Authenticator/HttpDigestAuthenticator.php]: Do we need the null alternative here? I would simply do `if (!$result)` to check for emptiness - the strict ones can use `$result === []`. The less return types the better, not?

File: src/Authenticator/TokenAuthenticator.php [hunk 2]
```diff
@@ -98,11 +98,11 @@ protected function getTokenFromQuery(ServerRequestInterface $request, $queryPara
     {
         $queryParams = $request->getQueryParams();
 
-        if (isset($queryParams[$queryParam])) {
-            return $queryParams[$queryParam];
+        if (!array_key_exists($queryParam, $queryParams)) {
```
Comment (dereuromark) [inline src/Authenticator/TokenAuthenticator.php]: why array_key_exists? It is either non-null or irrelevant IMO.

File: src/Authenticator/TokenAuthenticator.php [hunk 3]
```diff
@@ -98,11 +98,11 @@ protected function getTokenFromQuery(ServerRequestInterface $request, $queryPara
...
+            return null;
         }
 
-        return null;
+        return $queryParams[$queryParam];
```
Comment (dereuromark) [inline src/Authenticator/TokenAuthenticator.php]: :+1: on inverting.

File: src/Authenticator/HttpDigestAuthenticator.php (hunk 1 above)
Comment (inoas) [reply src/Authenticator/HttpDigestAuthenticator.php]: You had the same question on the original ticket:
https://github.com/cakephp/authentication/pull/175#pullrequestreview-84302555
Is the conclusion still the same?

File: src/Authenticator/TokenAuthenticator.php (hunk 2 above)
Comment (inoas) [reply src/Authenticator/TokenAuthenticator.php]: Well I thought it doesn't matter what's inside, null, 0, '0', '' - all ok, and middleware can do that (mine casts to real types for API things) a token with `(int)0` should be legit.

Let me ask the other way around: What's the benefit of using isset here which does all that autocasting? Does array_key_exists hurt?

File: src/Authenticator/HttpDigestAuthenticator.php (hunk 1 above)
Comment (dereuromark) [reply src/Authenticator/HttpDigestAuthenticator.php]: OK, sry, didnt remember that one.

File: src/Authenticator/TokenAuthenticator.php (hunk 2 above)
Comment (dereuromark) [reply src/Authenticator/TokenAuthenticator.php]: isset doesnt do any auto-cast. array_key_exists() usually goes a) too fur, b) is 10-100x slower.

Comment (inoas) [reply src/Authenticator/TokenAuthenticator.php]: I still want to see the performance difference on 7.2/7.1 but going to remove this now.

File: src/Authenticator/JwtAuthenticator.php [hunk 4]
```diff
@@ -126,11 +127,13 @@ public function getPayload(ServerRequestInterface $request = null)
         $payload = null;
         $token = $this->getToken($request);
 
-        if ($token) {
+        if ($token !== null) {
             $payload = $this->decodeToken($token);
         }
 
-        return $this->payload = $payload;
```
Comment (inoas) [inline src/Authenticator/JwtAuthenticator.php]: Is this good style?
If not, could it be caught by some checker in future?

Comment (dereuromark) [reply src/Authenticator/JwtAuthenticator.php]: It is not necessarily something of an issue.

Comment (markstory) [reply src/Authenticator/JwtAuthenticator.php]: The previous way uses 2 fewer lines of code, and is readable in my opinion. This form is fine too though.

Comment (inoas) [reply src/Authenticator/JwtAuthenticator.php]: One last thing that bugs me is that a getters modifies the instance property...

**Pull request merged**